### ISSUE TITLE
inconsistent image fingerprint on uk.images.linuxcontainers.org/1.0/images

### ISSUE THREAD

**Issue by erickeller** [lxc/linuxcontainers.org#302]
Hi there,

with @BrunnerLivio we notice some inconsistency in the images on the https://images.linuxcontainers.org server.

basically the image listed do have a consistent fingerprint when accessing the image details as depicted in the following code:

```
curl https://uk.images.linuxcontainers.org/1.0/images | python -m json.tool                                            
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 23352  100 23352    0     0  23352      0  0:00:01 --:--:--  0:00:01 91218
{
    "metadata": [
        "/1.0/images/17ba09371cc3443715d6ce8733c1cc226645863c139450eb503d8eb08a1f7250",
        "/1.0/images/e2a9549f1f318c8faa64963c67efa95ced8572368127666cf6760a8eb7ed0853",
        "/1.0/images/0c03fddf670309012e6bb7bd8545c6c0d6b725ffb84f2dab67d8542a44e3c50b",
        "/1.0/images/f79d48030270234045acba9239cba77dbd3a0ea6e77ab1ec741834dbbb58ee78",
        "/1.0/images/adb7164bbcdcf14a1be1dc8dfecd2825333e6d40757e053617ef9640f5e99fd5",
        "/1.0/images/2bc172159ce57bbb02a1372ad5016f8f3374c709a98760a1059405363424fb50",
        "/1.0/images/0c252ff1d64978e89052f81bc71c91912ed59521658c260facf9cdc1af1b0553",
        "/1.0/images/8884bce09c6b6e5ca573a833e5fb7872c1e3ef44807e9806bd693ccbc97f7d27",
        "/1.0/images/3453e9af2bec38845bb8afa393e6d3e3ba81d449d82ac1748c06683002b7a1c0",
        "/1.0/images/31b1f1fdc46d556963e84b3dd9d482be03916331730d159c89fb8ab129c082f0",
        "/1.0/images/bd847ed151fafbc8d5bf590f0a35cc3fbec87bbda79200ad15564e05d6fa97a6",
        "/1.0/images/faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",
       [...]
        "/1.0/images/c6e9b0dcca19764861773c7808d8cb472781009601402142034a818f52c48a55"
    ],
    "status": "Success",
    "status_code": 200,
    "type": "sync"
}
```
Getting some image details of the `/1.0/images/faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6` results in
the same fingerprint:
 "fingerprint": "faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",

```
curl https://uk.images.linuxcontainers.org/1.0/images/faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6 | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   790  100   790    0     0    790      0  0:00:01 --:--:--  0:00:01  1850
{
    "metadata": {
        "aliases": [
            {
                "description": "Alpine 3.5 (amd64) (default)",
                "name": "alpine/3.5/amd64/default",
                "target": "alpine/3.5/amd64/default"
            },
            {
                "description": "Alpine 3.5 (amd64)",
                "name": "alpine/3.5/amd64",
                "target": "alpine/3.5/amd64"
            }
        ],
        "architecture": "x86_64",
        "created_at": "2018-03-08T18:00:24Z",
        "expires_at": "1970-01-01T00:00:00Z",
        "filename": "alpine-3.5-amd64-default-20180308_17:50.tar.xz",
        "fingerprint": "faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",
        "properties": {
            "architecture": "amd64",
            "build": "20180308_17:50",
            "description": "Alpine 3.5 (amd64) (20180308_17:50)",
            "distribution": "alpine",
            "release": "3.5"
        },
        "public": true,
        "size": 1749138,
        "uploaded_at": "2018-03-08T18:00:24Z"
    },
    "status": "Success",
    "status_code": 200,
    "type": "sync"
}
```

Now comes the tricky part, using the lxc (lxc version 2.21) to copy this image from with this fingerprint locally fails:

```
lxc image copy images:faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6 local: --copy-aliases           
error: The requested image couldn't be found.
```
It seems that the exposed REST-API for this server is delivering erroneous/outdated data...

Using the lxc command to get some more information about the alias target image, results in a complete other image Fingerprint :(

```
lxc image info images:alpine/3.5/amd64                                                                                 
Fingerprint: 9677bdac722171ddc16f3a77c800f840d51e1c46ee7e627c5141d8b237220f5f
Size: 1.70MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/08 00:00 UTC
    Uploaded: 2018/03/08 00:00 UTC
    Expires: never
    Last used: never
Properties:
    serial: 20180308_17:50
    description: Alpine 3.5 amd64 (20180308_17:50)
    os: Alpine
    release: 3.5
    architecture: amd64
Aliases:
    - alpine/3.5/default
    - alpine/3.5/default/amd64
    - alpine/3.5
    - alpine/3.5/amd64
Cached: no
Auto update: disabled
``` 

After some digging we noticed that a really small amount of listed image correspond to the proper fingerprint:

```
lxc image list images: --format yaml | grep -i fingerprint | awk '{print $2}' > lxc-images.fingerprints
curl https://uk.images.linuxcontainers.org/1.0/images | python -m json.tool | grep -i images | cut -d '/' -f4 | tr -d '[",]' > linuxcontainers.org.fingerprints
# comparison
for im in `cat lxc-images.fingerprints`; do grep $im linuxcontainers.org.fingerprints; done                            
4cd85aceb09377aaaf0f67037fa9c00dee46d358012077174a4fe7c1f2b1f000
b1aeef2293c0bcc4639983a52e5eff47600361a27c4e2c5b0b3417362f68c44d
1fd0905297acbd9564b9556b22cc0fb6337c7f3c633ed9be94423bdefcb5bacb
a14a23a034aab22eb4ebbb5826bcb21862c41f4ad07aeb922dcdb68bc46d76e9
fc2de6fd2745b5e94a73ec48ea5c6c3391e82f555d2d9e9c06a0147637cf18e6
6fce0e9befe843135fc6e5c00dd2ef7158feba80cf0eb022b097d967d318d47d
```

Well the initial idea was to use the LXD rest-api to list the content of the https://images.linuxcontainers.org  server and to make some copy operation via webui over this API as specified in https://github.com/lxc/lxd/blob/master/doc/rest-api.md

`1.0/images/*`

Coud someone have a look at this issue?
Is there an API description to access the simplestreams protocol so it's consistent with the lxc command?

Basically we would need the complete list of images `1.0/images/*` and their details `1.0/images/<image_fingerprint>`

Cheers
Eric

**Comment by stgraber** [comment]
So the problem is that you're mixing the LXD and simplestreams protocols. As they serve different artifacts for the same images, the fingerprints are different.

For example:

```
stgraber@castiana:~$ lxc image info img-lxd:alpine/3.5/amd64
Fingerprint: faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6
Size: 1.67MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/08 18:00 UTC
    Uploaded: 2018/03/08 18:00 UTC
    Expires: never
    Last used: never
Properties:
    release: 3.5
    architecture: amd64
    description: Alpine 3.5 (amd64) (20180308_17:50)
    build: 20180308_17:50
    distribution: alpine
Aliases:
    - alpine/3.5/amd64/default (Alpine 3.5 (amd64) (default))
    - alpine/3.5/amd64 (Alpine 3.5 (amd64))
Cached: no
Auto update: disabled

stgraber@castiana:~$ lxc image info img-simple:alpine/3.5/amd64
Fingerprint: 9677bdac722171ddc16f3a77c800f840d51e1c46ee7e627c5141d8b237220f5f
Size: 1.70MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/08 00:00 UTC
    Uploaded: 2018/03/08 00:00 UTC
    Expires: never
    Last used: never
Properties:
    os: Alpine
    release: 3.5
    architecture: amd64
    serial: 20180308_17:50
    description: Alpine 3.5 amd64 (20180308_17:50)
Aliases:
    - alpine/3.5/default
    - alpine/3.5/default/amd64
    - alpine/3.5
    - alpine/3.5/amd64
Cached: no
Auto update: disabled
```

Those two are the same image but not packed the same way and since packing affects the files on disk, it affects the fingerprint.

The "LXD" protocol only returns split images that are made of a .tar.xz metadata tarball and a .tar.xz rootfs tarball. The "simplestreams" protocol has two versions of each of the images. It's got the same .tar.xz metadata + .tar.xz rootfs but it also has a .tar.xz metadata + .squashfs rootfs. That latter is what's returned for any client which supports squashfs.

The simplestreams entry for the image above is:

```
        "20180308_17:50": {
          "items": {
            "root.20180307_17:50.vcdiff": {
              "path": "images/alpine/3.5/amd64/default/20180308_17:50/delta-20180307_17:50.vcdiff",
              "sha256": "4c406b9bab06cae212be5ca99eb4aab6f9a2f0d9b86998458c77fa3e29f2222c",
              "ftype": "squashfs.vcdiff",
              "delta_base": "20180307_17:50",
              "size": 10026
            },
            "root.tar.xz": {
              "path": "images/alpine/3.5/amd64/default/20180308_17:50/rootfs.tar.xz",
              "sha256": "fe475fcf45a22257d513795a40b148b481e8e8aa376064f6e0495d4f45f3b5ee",
              "ftype": "root.tar.xz",
              "size": 1748188
            },
            "root.squashfs": {
              "path": "images/alpine/3.5/amd64/default/20180308_17:50/rootfs.squashfs",
              "sha256": "c7338523d1d0cb76b8159fa7b86cfde213e7febdf6f6e02f5d7dcc09346a6bb3",
              "ftype": "squashfs",
              "size": 1777664
            },
            "lxd.tar.xz": {
              "combined_squashfs_sha256": "9677bdac722171ddc16f3a77c800f840d51e1c46ee7e627c5141d8b237220f5f",
              "sha256": "a0eab2c6fb06e9dcca1dee3b21df357213d1d6c948abca39dfce41ced1ef031d",
              "path": "images/alpine/3.5/amd64/default/20180308_17:50/lxd.tar.xz",
              "ftype": "lxd.tar.xz",
              "combined_rootxz_sha256": "faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",
              "combined_sha256": "faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",
              "size": 596
            }
          }
        }
```

The key here being:

```
"combined_squashfs_sha256": "9677bdac722171ddc16f3a77c800f840d51e1c46ee7e627c5141d8b237220f5f",
"combined_rootxz_sha256": "faf04466a305b2a874284b5cd54ddd7513e6e734c02de1da71b4a53e016103f6",
```

Those are the LXD fingerprints, the first when using the squashfs, the second when using the tar.xz and they line up with what we see through the simplestreams and LXD protocols respectively.

**Comment by stgraber** [comment]
Oh, I should mention, the two remotes I'm using above to test this have been added with:

```
lxc remote add img-lxd https://images.linuxcontainers.org --protocol=lxd
lxc remote add img-simple https://images.linuxcontainers.org --protocol=simplestreams
```

These days, the "images" remote on all LXD clients should be using the "simplestreams" protocol. We switched the default a long time ago and have code that's supposed to automatically rewrite our users' config to switch from LXD to simplestreams on existing installs.

That said, as the "images" remote isn't a hardcoded one, individual users can change it for something else...

**Comment by erickeller** [comment]
Thanks for the clarification. Concerning the lxd rest-api calls to copy an image via specified lxd protocol from the https://images.linuxcontainers.org remote should then do the trick.

About the simplestreams could you give an hint about the path where it is exposed on the Linux containers.org https://us.images.linuxcontainers.org/releases/streams/v1/ did not work.

As far as I understand using the lxd REST-API just takes a protocol parameter in it's properties

```
# LXD protocol
lxc image copy images:6e2aaec2c4e4d615390fa4418e356095666e09b7a3062ab75dca802452f29b8c local: --debug --copy-aliases --public
DBUG[03-09|08:28:33] Sending request to LXD                   etag= method=POST url=http://unix.socket/1.0/images
DBUG[03-09|08:28:33] 
	{
		"auto_update": false,
		"properties": null,
		"public": true,
		"filename": "",
		"source": {
			"alias": "",
			"certificate": "",
			"protocol": "lxd",
			"server": "https://images.linuxcontainers.org:8443",
			"mode": "pull",
			"type": "image",
			"url": "",
			"name": "",
			"fingerprint": "6e2aaec2c4e4d615390fa4418e356095666e09b7a3062ab75dca802452f29b8c",
			"secret": ""
		},
		"compression_algorithm": "",
		"aliases": null
	}
```
```
# simplestreams protocol
lxc image copy ubuntu-daily:06d96148063f local: --public --copy-aliases --debug
	{
		"auto_update": false,
		"properties": null,
		"public": true,
		"filename": "",
		"source": {
			"alias": "",
			"certificate": "",
			"protocol": "simplestreams",
			"server": "https://cloud-images.ubuntu.com/daily",
			"mode": "pull",
			"type": "image",
			"url": "",
			"name": "",
			"fingerprint": "06d96148063f67e8328c5e91558369a3d03726a672fa47682cb3254ac33466c8",
			"secret": ""
		},
		"compression_algorithm": "",
		"aliases": null
	}
```

What's about the REST-API call for listing image using the different protocols:

```
# simplestreams protocol
lxc image list ubuntu-daily: 06d96148063f --debug                                                                                                                                                                                                                       
DBUG[03-09|08:44:45] Connecting to a remote simplestreams server 
+------------+--------------+--------+---------------------------------------------+--------+----------+------------------------------+
|   ALIAS    | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 |  ARCH  |   SIZE   |         UPLOAD DATE          |
+------------+--------------+--------+---------------------------------------------+--------+----------+------------------------------+
| b (7 more) | 06d96148063f | yes    | ubuntu 18.04 LTS amd64 (daily) (20180307.1) | x86_64 | 165.02MB | Mar 7, 2018 at 12:00am (UTC) |
+------------+--------------+--------+---------------------------------------------+--------+----------+------------------------------+
```
for the simple streams protocol is does not display any debug hint concerning the API call it does to get to this listing.

On the other hand using an image remote with the LXD protocol gives us much more information
especially getting all image in one recursion from the `https://images.linuxcontainers.org:8443/1.0/images?recursion=1` url

```
lxc image list images:6e2aaec2c4e4d615390fa4418e356095666e09b7a3062ab75dca802452f29b8c --debug                                                                                                                                                                          
DBUG[03-09|08:40:35] Connecting to a remote public LXD over HTTPs 
DBUG[03-09|08:40:35] Sending request to LXD                   etag= method=GET url=https://images.linuxcontainers.org:8443/1.0
DBUG[03-09|08:40:35] Got response struct from LXD 
DBUG[03-09|08:40:35] 
	{
		"config": null,
		"api_extensions": null,
		"api_status": "",
		"api_version": "",
		"auth": "guest",
		"public": true,
		"auth_methods": null,
		"environment": {
			"addresses": [
				"images.linuxcontainers.org:8443"
			],
			"architectures": null,
			"certificate": "",
			"certificate_fingerprint": "",
			"driver": "apache2",
			"driver_version": "",
			"kernel": "Linux",
			"kernel_architecture": "x86_64",
			"kernel_version": "",
			"server": "images.linuxcontainers.org",
			"server_pid": 0,
			"server_version": "1.0",
			"storage": "zfs",
			"storage_version": ""
		}
	} 
DBUG[03-09|08:40:35] Sending request to LXD                   etag= method=GET url="https://images.linuxcontainers.org:8443/1.0/images?recursion=1"
DBUG[03-09|08:40:36] Got response struct from LXD 
DBUG[03-09|08:40:36] 
	[
		{
			"auto_update": false,
			"properties": {
				"architecture": "amd64",
				"build": "20180306_17:50",
				"description": "Alpine 3.4 (amd64) (20180306_17:50)",
				"distribution": "alpine",
				"release": "3.4"
			},
			"public": true,
			"aliases": [],
			"architecture": "x86_64",
			"cached": false,
			"filename": "alpine-3.4-amd64-default-20180306_17:50.tar.xz",
			"fingerprint": "17ba09371cc3443715d6ce8733c1cc226645863c139450eb503d8eb08a1f7250",
			"size": 2096466,
			"created_at": "2018-03-06T18:00:09Z",
			"expires_at": "1970-01-01T00:00:00Z",
			"last_used_at": "0001-01-01T00:00:00Z",
			"uploaded_at": "2018-03-06T18:00:09Z"
		},
...
```

**Comment by stgraber** [comment]
Right, that's because simplestreams isn't a LXD protocol and isn't a REST API, so you don't see any request being made through the normal LXD communication stack.

Simplestreams pretty much just parses: https://images.linuxcontainers.org/streams/v1/images.json

**Comment by erickeller** [comment]
Last question about the properties of an image...
simplestream and LXD as you explained pack different metadata:

Simplestreams protocol
```
lxc remote add images-simplestreams https://images.linuxcontainers.org --protocol simplestreams --public
lxc image info images-simplestreams:ubuntu/xenial/amd64
Fingerprint: d1e45557fe3eda5457538542c73b6c07212b9ba80e17ae2e2c5f4b2e4a3ee350
Size: 93.32MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/09 00:00 UTC
    Uploaded: 2018/03/09 00:00 UTC
    Expires: never
    Last used: never
Properties:
    architecture: amd64
    serial: 20180309_03:49
    description: Ubuntu xenial amd64 (20180309_03:49)
    os: Ubuntu
    release: xenial
Aliases:
    - ubuntu/xenial/default
    - ubuntu/xenial/default/amd64
    - ubuntu/16.04/default
    - ubuntu/16.04/default/amd64
    - ubuntu/xenial
    - ubuntu/xenial/amd64
    - ubuntu/16.04
    - ubuntu/16.04/amd64
Cached: no
Auto update: disabled
```

LXD protocol
```
lxc remote add images-lxd https://images.linuxcontainers.org --protocol lxd --public
lxc image info images-lxd:ubuntu/xenial/amd64                                                                                                                                                                                                                           
Fingerprint: bb7901f3ddac26aeb0276f0f8a55c2f611abd91490de996db7b76a7099cfa894
Size: 83.64MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/09 04:23 UTC
    Uploaded: 2018/03/09 04:23 UTC
    Expires: never
    Last used: never
Properties:
    distribution: ubuntu
    release: xenial
    description: Ubuntu xenial (amd64) (20180309_03:49)
    build: 20180309_03:49
    architecture: amd64
Aliases:
    - ubuntu/xenial/amd64/default (Ubuntu xenial (amd64) (default))
    - ubuntu/xenial/amd64 (Ubuntu xenial (amd64))
Cached: no
Auto update: disabled
```

the properties of an image varies on the one hand there are called `distribution: ubuntu` and in the other hand `os: ubuntu`

Is there some normalisation planned?

**Comment by stgraber** [comment]
simplestreams properties are auto-generated on the client side, so we can't easily change things on the server side, for the LXD protocol we could make them match, but that'd break users that rely on the old names. What would probably be reasonable is to duplicate some of them so that using the simplestream names will work with both.

**Comment by stgraber** [comment]
There we go:

```
stgraber@castiana:~$ lxc image info img-simple:ubuntu/xenial/amd64
Fingerprint: d1e45557fe3eda5457538542c73b6c07212b9ba80e17ae2e2c5f4b2e4a3ee350
Size: 93.32MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/09 00:00 UTC
    Uploaded: 2018/03/09 00:00 UTC
    Expires: never
    Last used: never
Properties:
    architecture: amd64
    serial: 20180309_03:49
    description: Ubuntu xenial amd64 (20180309_03:49)
    os: Ubuntu
    release: xenial
Aliases:
    - ubuntu/xenial/default
    - ubuntu/xenial/default/amd64
    - ubuntu/16.04/default
    - ubuntu/16.04/default/amd64
    - ubuntu/xenial
    - ubuntu/xenial/amd64
    - ubuntu/16.04
    - ubuntu/16.04/amd64
Cached: no
Auto update: disabled
stgraber@castiana:~$ lxc image info img-lxd:ubuntu/xenial/amd64
Fingerprint: bb7901f3ddac26aeb0276f0f8a55c2f611abd91490de996db7b76a7099cfa894
Size: 83.64MB
Architecture: x86_64
Public: yes
Timestamps:
    Created: 2018/03/09 04:23 UTC
    Uploaded: 2018/03/09 04:23 UTC
    Expires: never
    Last used: never
Properties:
    distribution: ubuntu
    release: xenial
    os: ubuntu
    architecture: amd64
    description: Ubuntu xenial amd64 (20180309_03:49)
    build: 20180309_03:49
    serial: 20180309_03:49
Aliases:
    - ubuntu/xenial/amd64/default (Ubuntu xenial (amd64) (default))
    - ubuntu/xenial/amd64 (Ubuntu xenial (amd64))
Cached: no
Auto update: disabled
```

Note that it may take an hour or so for the mirror to get this change.

**Comment by erickeller** [comment]
nice! This is great :)
